// @framer/djfumberger.media-controls/code/App.tsx
import { Data, animate, Override, Animatable } from "framer"

// const data = Data({
//     play: false,
//     progress: 0.0,
//     elapsedTime: 0.0,
//     remainingTime: 0.0,
// })

// export const isProgressBar: Override = () => {
//     return {
//         progress: data.progress,
//     }
// }

// export const isElapsedTimeLabel: Override = () => {
//     return {
//         seconds: data.elapsedTime,
//     }
// }

// export const isRemainingTimeLabel: Override = () => {
//     return {
//         seconds: data.remainingTime,
//     }
// }

// export const isPlayer: Override = () => {
//     return {
//         shouldPlay: data.play,
//         debug: true,
//         onTimeUpdate(elapsed: number, duration: number) {
//             data.elapsedTime = Math.floor(elapsed)
//             data.remainingTime = duration - Math.floor(elapsed)
//         },
//         onProgress(progress: number) {
//             data.progress = progress
//         },
//         onEnded() {
//             data.play = false
//         },
//     }
// }

// export const isPlayButton: Override = () => {
//     return {
//         paused: data.play == false,
//         onPlay() {
//             data.play = true
//         },
//         onPause() {
//             data.play = false
//         },
//     }
// }

// export const isTapTarget: Override = () => {
//     return {
//         onTap() {
//             data.play = !data.play
//         },
//     }
// }
