// @framer/djfumberger.media-controls/code/CircularProgress.tsx
import * as React from "react"
import { PropertyControls, ControlType } from "framer"

type Props = {
    strokeWidth: number
    baseColor: string
    color: string
    progress: 0.5
    width: number
    height: number
}

export class ProgressCircle extends React.Component<Props> {
    style: React.CSSProperties = {
        height: "100%",
        display: "flex",
        alignItems: "center",
        justifyContent: "center",
        textAlign: "center",
        overflow: "hidden",
        pointerEvents: "none",
    }

    render() {
        let canvasWidth = this.props.width
        let canvasHeight = this.props.height
        let radius = canvasWidth / 2 - this.props.strokeWidth / 2
        let cx = radius + this.props.strokeWidth / 2
        let cy = radius + this.props.strokeWidth / 2

        let dashAmount = 2 * Math.PI * radius
        let dashOffset = dashAmount * (1.0 - this.props.progress)
        let amountStr = `${dashAmount} ${dashAmount}`

        let circleStyle = { ...this.style, transform: "rotate(-90deg)" }

        let baseCircle = (
            <circle
                r={radius}
                cx={cx}
                cy={cy}
                stroke={this.props.baseColor}
                strokeLinecap="round"
                strokeWidth={this.props.strokeWidth}
                fill="none"
            />
        )
        let circle = (
            <circle
                r={radius}
                cx={cx}
                cy={cy}
                stroke={this.props.color}
                strokeLinecap="round"
                strokeWidth={this.props.strokeWidth}
                fill="none"
                strokeDasharray={amountStr}
                strokeDashoffset={dashOffset}
            />
        )
        let svg = (
            <svg width={canvasWidth} height={canvasHeight}>
                {baseCircle}
                {circle}
            </svg>
        )
        return <div style={circleStyle}>{svg}</div>
    }

    static defaultProps: Props = {
        strokeWidth: 5,
        progress: 0.5,
        baseColor: "rgba(222,222,222,0.0)",
        color: "rgba(66,66,66,1.0)",
        width: 150,
        height: 150,
    }

    static propertyControls: PropertyControls<Props> = {
        color: { type: ControlType.Color, title: "Color" },
        baseColor: { type: ControlType.Color, title: "Base Color" },
        strokeWidth: {
            type: ControlType.Number,
            title: "Line Width",
            step: 0.1,
        },
        progress: {
            type: ControlType.Number,
            title: "Progress",
            min: 0,
            max: 1,
            step: 0.01,
        },
    }
}
